Add floating point matrix multiply-accumulate widening intrinsics

amilendra opened the pull request:
Adds intrinsic support for the FMMLA matrix multiply-add widening instructions introduced by the 2024 dpISA.

FEAT_F8F32MM: Neon/SVE2 FP8 to single-precision
FEAT_F8F16MM: Neon/SVE2 FP8 to half-precision
FEAT_SVE_F16F32MM: SVE half-precision to single-precision

Relands PR #409 that was approved, mistakenly merged and subsequently reverted.

---
name: Pull request
about: Technical issues, document format problems, bugs in scripts or feature proposal.

---

<!-- SPDX-FileCopyrightText: Copyright 2021-2022 Arm Limited and/or its affiliates <open-source-office@arm.com> -->
<!-- SPDX-License-Identifier: Apache-2.0 -->

**Thank you for submitting a pull request!**

If this PR is about a bugfix:

Please use the bugfix label and make sure to go through the checklist below.

If this PR is about a proposal:

We are looking forward to evaluate your proposal, and if possible to
make it part of the Arm C Language Extension (ACLE) specifications.

We would like to encourage you reading through the [contribution
guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md), in particular the section on [submitting
a proposal](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#proposals-for-new-content).

Please use the proposal label.

As for any pull request, please make sure to go through the below
checklist.

Checklist: (mark with ``X`` those which apply)

* [X] If an issue reporting the bug exists, I have mentioned it in the
      PR (do not bother creating the issue if all you want to do is
      fixing the bug yourself).
* [X] I have added/updated the `SPDX-FileCopyrightText` lines on top
      of any file I have edited. Format is `SPDX-FileCopyrightText:
      Copyright {year} {entity or name} <{contact informations}>`
      (Please update existing copyright lines if applicable. You can
      specify year ranges with hyphen , as in `2017-2019`, and use
      commas to separate gaps, as in `2018-2020, 2022`).
* [X] I have updated the `Copyright` section of the sources of the
      specification I have edited (this will show up in the text
      rendered in the PDF and other output format supported). The
      format is the same described in the previous item.
* [X] I have run the CI scripts (if applicable, as they might be
      tricky to set up on non-*nix machines). The sequence can be
      found in the [contribution
      guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration). Don't
      worry if you cannot run these scripts on your machine, your
      patch will be automatically checked in the Actions of the pull
      request.
* [X] I have added an item that describes the changes I have
      introduced in this PR in the section **Changes for next
      release** of the section **Change Control**/**Document history**
      of the document. Create **Changes for next release** if it does
      not exist. Notice that changes that are not modifying the
      content and rendering of the specifications (both HTML and PDF)
      do not need to be listed.
* [X] When modifying content and/or its rendering, I have checked the
      correctness of the result in the PDF output (please refer to the
      instructions on [how to build the PDFs
      locally](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration)).
* [X] The variable `draftversion` is set to `true` in the YAML header
      of the sources of the specifications I have modified.
* [] Please *DO NOT* add my GitHub profile to the list of contributors
      in the [README](https://github.com/ARM-software/acle/blob/main/README.md#contributors-) page of the project.

amilendra committed:
Add floating point matrix multiply-accumulate widening intrinsics

Adds intrinsic support for the FMMLA matrix multiply-add widening instructions
introduced by the 2024 dpISA.

FEAT_F8F32MM: Neon/SVE2 FP8 to single-precision
FEAT_F8F16MM: Neon/SVE2 FP8 to half-precision
FEAT_SVE_F16F32MM: SVE half-precision to single-precision

Lukacma commented on tools/intrinsic_db/advsimd.csv:
```diff
@@ -4810,3 +4810,7 @@ float32x4_t vmlalltbq_lane_f32_mf8_fpm(float32x4_t vd, mfloat8x16_t vn, mfloat8x
 float32x4_t vmlallttq_laneq_f32_mf8_fpm(float32x4_t vd, mfloat8x16_t vn, mfloat8x16_t vm, __builtin_constant_p(lane), fpm_t fpm)	vd -> Vd.4S;vm -> Vn.16B; vm -> Vm.B; 0 <= lane <= 15	FMLALLBB Vd.4S, Vn.16B, Vm.B[lane]	Vd.4S -> result	A64
+
+<SECTION>	Matrix multiplication intrinsics from Armv9.6-A
+float16x4_t vmmlaq_f16_mf8(float16x4_t r, mfloat8x16_t a, mfloat8x16_t b)	r -> Vd.4H;a -> Vn.16B;b -> Vm.16B	FMMLA Vd.4H, Vn.16B, Vm.16B	Vd.4H -> result	A64
```
The type for destination and first sourc should also be float16x8

Lukacma replied:
I think you might also be missing intrinsics for f16 to f32 variant. Or is that a separate patch ? https://developer.arm.com/documentation/ddi0602/2025-09/SIMD-FP-Instructions/FMMLA--widening--half-precision-to-single-precision---Half-precision-matrix-multiply-accumulate-to-single-precision-

amilendra committed:
Address review comments on the Neon FMMLA intrinsics

Add fpm source parameter
Change destination and first source parameter type: float16x4 -> float16x8

amilendra commented on tools/intrinsic_db/advsimd.csv:
```diff
@@ -4810,3 +4810,7 @@ float32x4_t vmlalltbq_lane_f32_mf8_fpm(float32x4_t vd, mfloat8x16_t vn, mfloat8x
 float32x4_t vmlallttq_laneq_f32_mf8_fpm(float32x4_t vd, mfloat8x16_t vn, mfloat8x16_t vm, __builtin_constant_p(lane), fpm_t fpm)	vd -> Vd.4S;vm -> Vn.16B; vm -> Vm.B; 0 <= lane <= 15	FMLALLBB Vd.4S, Vn.16B, Vm.B[lane]	Vd.4S -> result	A64
+
+<SECTION>	Matrix multiplication intrinsics from Armv9.6-A
+float16x8_t vmmlaq_f16_mf8(float16x8_t r, mfloat8x16_t a, mfloat8x16_t b, fpm_t fpm)	r -> Vd.4H;a -> Vn.16B;b -> Vm.16B	FMMLA Vd.4H, Vn.16B, Vm.16B	Vd.4H -> result	A64
```
Feel free to fix it. Thanks

AlfieRichardsArm commented:
See #420 for the fix

sallyarmneale reviewed (commented):
All good, no updates required
